packages/core/echo/echo-db/src/core-db/automerge-repo-replacement/doc-handle.ts

```diff
@@ -0,0 +1,138 @@
+      // e.g. a primitive, an array, etc. - it must be an object). But adding that constraint creates
+      // a bunch of other problems elsewhere so for now we'll just cast it here to make Automerge happy.
+      this._doc = A.from(options.initialValue as Record<string, unknown>) as T;
+      this._doc = A.emptyChange<T>(this._doc);
+      this._ready.wake();
+    } else {
+      this._doc = A.init<T>();
+    }
+  }
+
+  get url() {
+    return stringifyAutomergeUrl(this._documentId);
+  }
+
+  get documentId(): DocumentId {
+    return this._documentId;
+  }
+
+  get state() {
+    return this._ready.state === TriggerState.RESOLVED ? 'ready' : 'pending';
+  }
+
+  docSync(): A.Doc<T> {
+    return this._doc;
+  }
+
+  async doc(): Promise<A.Doc<T>> {
+    await this._ready.wait();
+    return this._doc;
+  }
+
+  async whenReady(): Promise<void> {
+    await this._ready.wait();
+  }
+
+  isReady(): boolean {
+    return this._ready.state === TriggerState.RESOLVED;
+  }
+
+  change(fn: (doc: A.Doc<T>) => void, opts?: A.ChangeOptions<any>): void {
+    invariant(this._doc, 'DocHandleReplacement.change called on deleted doc');
+    const headsBefore = A.getHeads(this._doc);
+    this._doc = opts ? A.change(this._doc, opts, fn) : A.change(this._doc, fn);
+    this.emit('change', {
+      handle: this,
+      doc: this._doc,
+      patches: A.diff(this._doc, headsBefore, A.getHeads(this._doc)),
+    });
+  }
+
+  changeAt(heads: A.Heads, fn: (doc: A.Doc<T>) => void, opts?: A.ChangeOptions<any>): Heads | undefined {
+    invariant(this._doc, 'DocHandleReplacement.changeAt called on deleted doc');
+    const headsBefore = A.getHeads(this._doc);
+    const { newDoc, newHeads } = opts ? A.changeAt(this._doc, heads, opts, fn) : A.changeAt(this._doc, heads, fn);
+
+    this._doc = newDoc;
+    this.emit('change', {
+      handle: this,
+      doc: this._doc,
+      patches: newHeads ? A.diff(this._doc, headsBefore, newHeads) : [],
+    });
+    return newHeads || undefined;
+  }
+
+  delete(): void {
+    this._callbacks?.onDelete();
+    this.emit('delete', { handle: this });
+    this._doc = undefined as any as Doc<T>;
+  }
+
+  /**
+   * @internal
+   */
+  _getLastWriteMutation(): Uint8Array | undefined {
+    invariant(this._doc, 'Doc is deleted, cannot get last write mutation');
+    if (A.getHeads(this._doc).join('') === this._lastSyncedHeads.join('')) {
+      return;
+    }
+
+    const mutation = A.saveSince(this._doc, this._lastSyncedHeads);
+    if (mutation.length === 0) {
+      return;
+    }
+    this._lastSyncedHeads = A.getHeads(this._doc);
+    return mutation;
+  }
+
+  /**
+   * @internal
+   */
+  _incrementalUpdate(mutation: Uint8Array) {
+    invariant(this._doc, 'Doc is deleted, cannot write mutation');
+    const headsBefore = A.getHeads(this._doc);
+
+    this._doc = A.loadIncremental(this._doc, mutation);
+    this._lastSyncedHeads = A.getHeads(this._doc);
```

Comment:
is there a potential race condition here: doc might have unsynced changes before `_incrementalUpdate` is called, this will set `_lastSyncedHeads` to include those changes even though they were never synced